staging/src/k8s.io/kubectl/pkg/cmd/create/create_service_test.go

```diff
@@ -17,17 +17,285 @@ limitations under the License.
+						{Name: "123", Protocol: "TCP", Port: 123, TargetPort: intstr.IntOrString{Type: 0, IntVal: 123, StrVal: ""}, NodePort: 0},
+						{Name: "234-1234", Protocol: "TCP", Port: 234, TargetPort: intstr.IntOrString{Type: 0, IntVal: 1234, StrVal: ""}, NodePort: 0},
+					},
+					Selector:  map[string]string{"app": "validate-ok"},
+					ClusterIP: "", ExternalIPs: []string(nil), LoadBalancerIP: ""},
+			},
+			expectErr: false,
+		},
+		{
+			name:        "invalid-ClusterIPNone",
+			serviceType: v1.ServiceTypeNodePort,
+			tcp:         []string{"123", "234:1234"},
+			clusterip:   v1.ClusterIPNone,
+			expectErr:   true,
+		},
+		{
+			name:        "TCP-none",
+			serviceType: v1.ServiceTypeClusterIP,
+			clusterip:   "",
+			expectErr:   true,
+		},
+		{
+			name:         "invalid-ExternalName",
+			serviceType:  v1.ServiceTypeExternalName,
+			tcp:          []string{"123", "234:1234"},
+			clusterip:    "",
+			externalName: "@oi:test",
+			expectErr:    true,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			options := ServiceOptions{
+				Name:      tc.name,
+				Type:      tc.serviceType,
+				TCP:       tc.tcp,
+				ClusterIP: tc.clusterip,
+			}
+
+			var service *v1.Service
+
+			err := options.Validate()
+			if err == nil {
+				service, err = options.createService()
+			}
+			if tc.expectErr && err == nil {
+				t.Errorf("%s: expected an error, but createService passes.", tc.name)
+			}
+			if !tc.expectErr && err != nil {
+				t.Errorf("%s: unexpected error: %v", tc.name, err)
+			}
+			if !apiequality.Semantic.DeepEqual(service, tc.expected) {
+				t.Errorf("%s: expected:\n%#v\ngot:\n%#v", tc.name, tc.expected, service)
+			}
+		})
+	}
+}
+
+func TestCreateServiceClusterIP(t *testing.T) {
+	tests := []struct {
+		name           string
+		tcp            string
+		service        *v1.Service
+		expectedOutput string
+	}{
+		{
+			name: "clusterip-ok",
+			tcp:  "8080:8000",
+			service: &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "my-service",
+					Labels: map[string]string{"app": "my-service"},
+				},
+				Spec: v1.ServiceSpec{
+					Type:     v1.ServiceType(v1.ServiceTypeClusterIP),
+					Selector: map[string]string{"app": "my-service"},
+					Ports: []v1.ServicePort{
+						{
+							Name:       "8080-8000",
+							Protocol:   "TCP",
+							Port:       8080,
+							TargetPort: intstr.IntOrString{Type: 0, IntVal: 8000},
+						},
+					},
+				},
+			},
+			expectedOutput: "service/my-service\n",
+		},
+	}
+	for _, test := range tests {
+
+		tf := cmdtesting.NewTestFactory().WithNamespace("test")
+		defer tf.Cleanup()
+
+		codec := scheme.Codecs.LegacyCodec(scheme.Scheme.PrioritizedVersionsAllGroups()...)
+
+		ns := scheme.Codecs.WithoutConversion()
+		info, _ := runtime.SerializerInfoForMediaType(ns.SupportedMediaTypes(), runtime.ContentTypeJSON)
+		decoder := ns.DecoderToVersion(info.Serializer, schema.GroupVersion{Group: "", Version: "v1"})
+
+		tf.Client = &fake.RESTClient{
```

Comment:
You should re-use the simplicity of verifying only the necessary bits. One of the goals was to simplify testing and not needing to fake the entire factory here.

Reply:
Do you have an example of a kubectl command that already uses such a simplified test case, so I know what I need to rewrite?

Reply:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/create/create_job_test.go is a collection of good examples.

Reply:
Ok, i hope i got it right now:
I basically just need to test the "important" functions: And those are "Validate" and "createService", correct?
At least this is what I did now.

And for the other 3 tests i had, i created equivalent test cases in "TestCreateServices" so test case coverage should still be good.